src/Interpreters/ApplyWithSubqueryVisitor.h

```diff
@@ -0,0 +1,30 @@
+#pragma once
+
+#include <map>
+
+#include <Parsers/IAST.h>
+
+namespace DB
+{
+// TODO After we support `union_with_global`, this visitor should also be extended to match ASTSelectQueryWithUnion.
```

Comment:
BTW it will be something like `with_global` because it will apply for all subqueries with correct scoping:

```
WITH 1 AS x SELECT x;
WITH 1 AS x SELECT * FROM (SELECT x);
WITH 1 AS x SELECT *, x FROM (WITH 2 AS x SELECT x AS y);
```